app/messages/emq/validator.rb

```diff
@@ -0,0 +1,92 @@
+# frozen_string_literal: true
+
+require 'net/http' # Add this line to require Net::HTTP
+require 'avro'
+require 'fileutils'
+
+module Emq
+  # This class should be responsible for validating messages against an Avro schema
+  # stored in RedPanda registry
+  class Validator
+    attr_reader :schema_config, :validate_obj
+
+    # Initialize the validator with the subject, version and registry URL
+    # @param [String] subject the subject of the schema
+    # @param [String] version the version of the schema
+    # @param [String] registry_url the URL of the schema registry
+    def initialize(subject, version, registry_url)
+      @subject = subject
+      @version = version
+      @registry_url = registry_url
+    end
+
+    # Validate a message against the schema
+    # @param [Hash] message the message to validate
+    def validate_message(message)
+      encode_message(message, create_message_schema)
+    rescue StandardError => e
+      Rails.logger.error("Error validating volume tracking message: <#{e.message}>")
+      raise
+    end
+
+    private
+
+    # Encode a message using the schema
+    # @param [Hash] message the message to encode
+    # @param [String] schema the schema to use for encoding
+    # @return [String] the encoded message
+    def encode_message(message, schema) # rubocop:disable Metrics/MethodLength
+      begin
+        schema = Avro::Schema.parse(schema)
+      rescue Avro::SchemaParseError => e
+        Rails.logger.error("Schema parsing error: <#{e.message}>. Schema: #{schema}")
+        raise
+      end
+      stream = StringIO.new
+      writer = Avro::IO::DatumWriter.new(schema)
+      encoder = Avro::IO::BinaryEncoder.new(stream)
+      encoder.write("\xC3\x01") # Avro single-object container file header
+      encoder.write([schema.crc_64_avro_fingerprint].pack('Q')) # 8 byte schema fingerprint
+      writer.write(message, encoder)
+      stream.string
+    end
+
+    # Create the message schema
+    # @return [String] the schema for the message
+    def create_message_schema
+      # Prefer to use the cached schema if it exists.
+      cache_file_path = "data/avro_schema_cache/#{@subject}_v#{@version}.avsc"
```

Comment:
Using a cached schema is really nice. Less networking overhead :)

Reply:
Credit goes to @sdjmchattie as I used it in the same way as he did in Sequencescape. :)